### PR TITLE
[VarDumper] Output the location of calls to dump() 

### PR DESCRIPTION
**Pull request by ktherage**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | yes
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | not tested yet
| Fixed tickets | #30830
| License       | MIT
| Doc PR        |  

see #30830

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
The implementation relies on a statefull service. This should be avoided.
If we need dumping some context, it should be added to the `Data` object to me.
Then, dumpers could read it and do things appropriately.
See also `SourceContextProvider` which has move intensive logic for extracting the stack trace.

**Comment by ktherage** [comment]
@nicolas-grekas Thank you for the feed back. I've changed the implementation.

Do you have an idea on how can i insulate this new feature and set her only for CliDumper?

Because currently, without forcing file and line to `null` in `getDump` method of `VarDumperTestTrait` it has impact on all other tests and so maybe on functionnalities.

**Comment by Simperfit** [comment]
@ktherage can we move forward on this ;) ?

**Comment by ktherage** [comment]
@Simperfit I'm stuck on unit tests. I this they are to much nested each other. Testing cli dumper makes HtmlDumper, ServerDumper and all the caster crashed.

**Comment by ktherage** [comment]
@Simperfit @nicolas-grekas, I did some work on that, tests are finally green.
Could you, please, take a sometime to check those changes?

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php [hunk 1]
```diff
@@ -83,6 +84,8 @@ public function __construct($output = null, string $charset = null, int $flags =
             ]);
         }
 
+        $this->handlesHrefGracefully = 'JetBrains-JediTerm' !== getenv('TERMINAL_EMULATOR');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: why moving this up? if not needed, it should be reverted
I suppose you forgot to keep the check for KONSOLE_VERSION? (see removed line)

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php [hunk 2]
```diff
@@ -632,10 +631,29 @@ private function isWindowsTrueColor()
...
+        }
+
+        return '';
+    }
+
+    private function getSourceLink($file, $lineNumber)
     {
         if ($fmt = $this->displayOptions['fileLinkFormat']) {
-            return \is_string($fmt) ? strtr($fmt, ['%f' => $file, '%l' => $line]) : ($fmt->format($file, $line) ?: 'file://'.$file);
+            return \is_string($fmt) ? strtr($fmt, ['%f' => $file, '%l' => $lineNumber]) : ($fmt->format($file, $lineNumber) ?: 'file://'.$file);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: please revert unrelated changes

File: src/Symfony/Component/VarDumper/Dumper/ContextProvider/CliContextProvider.php [hunk 3]
```diff
@@ -24,9 +24,15 @@ public function getContext(): ?array
             return null;
         }
 
+        ['file' => $file, 'line' => $line] = (new SourceContextProvider())->getContext();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/ContextProvider/CliContextProvider.php]: each context provider should do one thing, this cannot be the correct design

File: src/Symfony/Component/VarDumper/Tests/Dumper/CliDumperTest.php [hunk 4]
```diff
@@ -204,7 +211,7 @@ public function testXmlResource()
         $var = xml_parser_create();
 
         $this->assertDumpMatchesFormat(
-            <<<'EOTXT'
+            <<<EOTXT
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Tests/Dumper/CliDumperTest.php]: please revert unrelated changes

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php [hunk 5]
```diff
@@ -491,7 +490,7 @@ protected function style($style, $value, $attr = [])
                 }
             }
             if (isset($attr['href'])) {
-                $value = "\033]8;;{$attr['href']}\033\\{$value}\033]8;;\033\\";
+                $value .= " \033]8;;{$attr['href']}\033\\[^]\033]8;;\033\\";
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: this breaks the case where the value should be the text of the link

File: src/Symfony/Component/VarDumper/Dumper/CliDumper.php [hunk 6]
```diff
@@ -619,7 +618,7 @@ private function isWindowsTrueColor()
             || 'xterm' === getenv('TERM')
             || 'Hyper' === getenv('TERM_PROGRAM');
 
-        if (!$result) {
+        if (!$result && \PHP_VERSION_ID >= 70200) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/VarDumper/Dumper/CliDumper.php]: why? if confirmed, this should be submitted separately.

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
I'm sorry but I'm not comfortable at all with the way this is implemented. This feels like breaking all semantics to fit the target.
- Data::dump() is about the represented structure. It cannot snif the context.
- CliDumper is about dumping the same structure, not the context.

 I think ServerDumper does it properly:
- the context is collected completly separately, in a flexible way
- the context is dumped separately

Here is a proposal: add a ContextualizedDumper that collects+dumps contexts and dumps using a decorated dumper - very alike ServerDumper. Writting to the output should be done using the decorated dumper too (unlike ServerDumpCommand, which write to Console i/o)

I hope this helps.

**Comment by ktherage** [comment]
@nicolas-grekas I've applied your recommendations (hoping i've not misread them). 
Can you review them please ?

**Comment by nicolas-grekas** [comment]
I push-forced on your fork to rebase+target 4.4, and do some changes.
Can you please try if that works for you?
Also, I think this doesn't work when dump is called from inside console commands.
Can you also check that?

**Comment by ktherage** [comment]
> I push-forced on your fork to rebase+target 4.4, and do some changes.
> Can you please try if that works for you?
> Also, I think this doesn't work when dump is called from inside console commands.
> Can you also check that?

@nicolas-grekas i've adapter unit test according to your changes and i've added one test for console command. Can you check them and tell me if they are good for you?

**Comment by ktherage** [comment]
As we talked together, i've adapted tests to a smaller sample an removed the fix i made on CliDumper for tests. 

I've also tried to call dump in a command of a website-skeleton and, as you pre-minded it, it does not worked but i don't know why (see the GIF bellow).

Peek 18-07-2019 17-03

**Comment by nicolas-grekas** [comment]
@ktherage the reason is that we wire a service as a dumper. Looks for cli_dumper/html_dumper I think in the code base

**Comment by ktherage** [comment]
> @ktherage the reason is that we wire a service as a dumper. Look for cli_dumper/html_dumper I think in the code base

Yes, [here it is](https://github.com/symfony/symfony/blob/4.4/src/Symfony/Bundle/DebugBundle/Resources/config/services.xml#L37). 
Do you mean that i shall add a service definition like that one below :
```xml
        <service id="var_dumper.contextualized_cli_dumper" class="Symfony\Component\VarDumper\Dumper\ContextualizedDumper">
            <argument type="service" id="var_dumper.cli_dumper" />
            <argument type="collection">
                <argument type="service" key="source">
                    <service class="Symfony\Component\VarDumper\Dumper\ContextProvider\SourceContextProvider">
                        <argument>%kernel.charset%</argument>
                        <argument type="string">%kernel.project_dir%</argument>
                        <argument type="service" id="debug.file_link_formatter" on-invalid="null" />
                    </service>
                </argument>
            </argument>
        </service>
```
and use this new service in the debug bundle in place of the `var_dumper.cli_dumper` service ?

**Comment by ktherage** [comment]
@nicolas-grekas I've worked on that and fixed the error for commands (see bellow).

I also tried to fixe the tests. But for me the CI behave strangely (depending on the php version).

Peek 10-09-2019 15-05

**Review by nicolas-grekas — APPROVED** [review]
(I tested locally and fixed tests)

**Comment by nicolas-grekas** [comment]
Thank you @ktherage.

**Comment by alimranahmed** [comment]
> Peek 10-09-2019 15-05

It doesn't work this way in browser.

**Pull request merged**